File: Documentation/3.5/Samples/java/src/test/java/net/ravendb/clientapi/session/configuration/OptimisticConcurrency.java
package net.ravendb.clientapi.session.configuration;

import net.ravendb.abstractions.basic.CleanCloseable;
import net.ravendb.client.IDocumentSession;
import net.ravendb.client.IDocumentStore;
import net.ravendb.client.document.DocumentSession;
import net.ravendb.client.document.DocumentStore;
import net.ravendb.samples.northwind.Product;


public class OptimisticConcurrency {

  public OptimisticConcurrency() throws Exception {
    try (IDocumentStore store = new DocumentStore()) {
      //region optimistic_concurrency_1
      try (IDocumentSession session = store.openSession()) {
        session.advanced().setUseOptimisticConcurrency(true);

        Product product = new Product();
        product.setName("Some Name");

        session.store(product, "products/999");
        session.saveChanges();

        try (IDocumentSession otherSession = store.openSession()) {
          Product otherProduct = otherSession.load(Product.class, "products/999");
          otherProduct.setName("Other name");

          otherSession.saveChanges();
        }

        product.setName("Better Name");
        session.saveChanges(); // throws ConcurrencyException
      }
      //endregion

      //region optimistic_concurrency_2
      store.getConventions().setDefaultUseOptimisticConcurrency(true);

      try (IDocumentSession session = store.openSession()) {
        boolean isSessionUsingOptimisticConcurrency = session.advanced().isUseOptimisticConcurrency(); // will return true
      }
      //endregion

      //region optimistic_concurrency_3
      store.getConventions().setDefaultUseOptimisticConcurrency(true);

      try (DocumentSession session = (DocumentSession) store.openSession()) {
        try (CleanCloseable _ = session.getDatabaseCommands().forceReadFromMaster()) {
          // In replicated setup where ONLY reads are load balanced (FailoverBehavior.ReadFromAllServers)
          // and optimistic concurrency checks are turned on
          // you must set 'ForceReadFromMaster' to get the appropriate ETag value for the document
          // when you want to perform document updates (writes)
          // because writes will go to the master server and ETag values between servers are not synchronized

          Product product = session.load(Product.class, "products/999");
          product.setName("New Name");

          session.saveChanges();
        }
      }
      //endregion
    }
  }
}

File: Documentation/3.5/Samples/csharp/Raven.Documentation.Samples/ClientApi/Session/Configuration/OptimisticConcurrency.cs
using Raven.Client;
using Raven.Client.Document;
using Raven.Documentation.CodeSamples.Orders;

namespace Raven.Documentation.Samples.ClientApi.Session.Configuration
{
	public class OptimisticConcurrency
	{
		public OptimisticConcurrency()
		{
			using (var store = new DocumentStore())
			{
				#region optimistic_concurrency_1
				using (IDocumentSession session = store.OpenSession())
				{
					session.Advanced.UseOptimisticConcurrency = true;

					Product product = new Product { Name = "Some Name" };

					session.Store(product, "products/999");
					session.SaveChanges();

					using (IDocumentSession otherSession = store.OpenSession())
					{
						Product otherProduct = otherSession.Load<Product>("products/999");
						otherProduct.Name = "Other Name";

						otherSession.SaveChanges();
					}

					product.Name = "Better Name";
					session.SaveChanges(); // will throw ConcurrencyException
				}
				#endregion

				#region optimistic_concurrency_2
				store.Conventions.DefaultUseOptimisticConcurrency = true;

				using (IDocumentSession session = store.OpenSession())
				{
					bool isSessionUsingOptimisticConcurrency = session.Advanced.UseOptimisticConcurrency; // will return true
				}
                #endregion

                #region optimistic_concurrency_3
                store.Conventions.DefaultUseOptimisticConcurrency = true;

                using (DocumentSession session = (DocumentSession)store.OpenSession())
                using (session.DatabaseCommands.ForceReadFromMaster())
                {
                    // In replicated setup where ONLY reads are load balanced (FailoverBehavior.ReadFromAllServers)
                    // and optimistic concurrency checks are turned on
                    // you must set 'ForceReadFromMaster' to get the appropriate ETag value for the document
                    // when you want to perform document updates (writes)
                    // because writes will go to the master server and ETag values between servers are not synchronized

                    Product product = session.Load<Product>("products/999");
                    product.Name = "New Name";

                    session.SaveChanges();
                }
                #endregion
            }
		}
	}
}